### ISSUE TITLE
PHP7, missing return types declaration (optional)

### ISSUE THREAD

**Issue by kalessil** [kalessil/phpinspectionsea#28]
Original link: https://bitbucket.org/kalessil/phpinspectionsea/issues/285/suggestion-weak-warning-php-7-return-type
Author: @mfn

See original ticket

**Comment by kalessil** [comment]
@mfn : is this ticket still relevant (cleaning tickets up) - JetBrains were working hard in this direction last half-year?

**Comment by mfn** [comment]
Jup, still relevant. There's no existing inspection which specifically what the title of this issue is.

We got other stuff, like "Missing return tag" or "Incompatible return type" but not something like "Missing return type declaration" or whatever we want to call it.

**Comment by funivan** [comment]
We can implement even fixer to set specific return type. 
And with php 7.1. we get nullable return types and `void`

**Comment by kalessil** [comment]
I'd actually target types from both 7.1/7.0 and analyzed all return statements in functions and methods.

**Comment by kalessil** [comment]
So, requirements engineering:

- inspection applied to non-abstract and non-magic methods from classes/traits;
- PHP language level is taken into account;
- method DocBlock must be completely ignored, types are resolved natively; with QF supporting following:
  - nullable return types identification;
  - void return type identification;

@mfn: did I miss anything here?

**Comment by mfn** [comment]
> did I miss anything here?

No expert 😄 but sounds reasonable to me.

> inspection applied to non-abstract

I understand for non-magic methods but abstract .. might be useful(?), but not strong opinion about it.

**Comment by kalessil** [comment]
Let give a try, abstracts return type can be taken from DocBlock directly - easy to pick in fact.

**Comment by kalessil** [comment]
Status update:
- implemented with QickFix (QF declares fully qualified classes, not sure if this should be similar to ::class QF);
- tests needs to be added;

If anyone wants to play with: fresh binary pushed as well.

**Comment by funivan** [comment]
@kalessil install plugin from binary.

**Comment by funivan** [comment]
Notice. It will be cool to have option in the configuration panel  "Import classes automatically"
I know that this is not so simple to do, but ti is very annoying to do it manually)

**Comment by kalessil** [comment]
Heh, seems like @funivan got addicted to quick-fixes =)

**Comment by funivan** [comment]
False positive:
```php
 class A  {
    public function run(Request $request = null) {
      if (null !== $request->get('do')) {
        return new RedirectResponse('/');
      }
    }
  }
```

In this case we will break our code if we add `RedirectResponse` to the return type

**Comment by rentalhost** [comment]
For PHP 7.1, maybe it could uses a nullable type: `?RedirectResponse`.

**Comment by kalessil** [comment]
PhpDoc is missing, no implicit null return - no chance to suggest nullable here.

**Comment by rentalhost** [comment]
I guess that for this case, when you don't have `return` as last statement, then you could consider that it coulds returns `null` too.

**Comment by kalessil** [comment]
Nope, that's not a feasible assumption - it can be throw exception or something else.

**Comment by kalessil** [comment]
`method.getType().global(holder.getProject()).filterUnknown().getTypes()` - that's how types got resolved, and missing null in resolved types totally makes sence.

**Comment by rentalhost** [comment]
```php
function x () {
    if ($this->error) {
        throw new ErrorException;
    }

    if ($this->good) {
        return new Good;
    }
}
```

In this case, it still should returns `?Good`, even if exception could happen (it doesn't affects the return typehint). If you have sure that you will always return `Good`, then `if()` is needless here, or you should create a new "generic" return that returns the `Good` too.

There are some real example that you have thought that can causes a problem?

**Comment by kalessil** [comment]
I guess I have to check the last statement as suggested @rentalhost ... x_x

**Comment by kalessil** [comment]
And I should admit that in spite of my analytical skills, I'd come with this solution fast, so thank you @rentalhost

**Comment by rentalhost** [comment]
That is the advantage of open source world. We can "combine" our minds to improve application.

**Comment by kalessil** [comment]
And the fix is pushed with fresh binary, next feedback round?

**Comment by kalessil** [comment]
Pilot version of the inspection, QF enhancement follow-up: #243

**Comment by funivan** [comment]
I will check new version tomorrow

**Comment by kalessil** [comment]
Super, thank you @funivan
